### PR TITLE
Implement unit tests for ChildProcessChannel - Closes #3117 

### LINKED ISSUE #3117: Unit test ChildProcessChannel class
### Expected behavior
Expect to `ChildProcessChannel` class to be covered by unit tests using Jest framework

### Actual behavior
It doesn't

### Which version(s) does this affect? (Environment, OS, etc...)
`feature/child_process_module`

### PR DESCRIPTION
**Pull request by limiaspasdaniel**
### What was the problem?
There were no **unit tests** implemented for `ChildProcessChannel`
### How did I fix it?
By implementing **unit tests** for `ChildProcessChannel` using **Jest Framework**
### How to test it?
`npm run jest:unit child_process_channel`

### Review checklist
* The PR resolves #3117 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+		call: jest.fn(),
+	};
+};
+axonRpc.Server = function() {
+	return {
+		expose: jest.fn(),
+	};
+};
+
+const EventEmitter2 = require('eventemitter2');
```
Comment (shuse2) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: import before mocking?

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+		moduleAlias: originalModuleAlias,
+		events: ['event1', 'event2'],
+		actions: {
+			action1: jest.fn(),
+			action2: jest.fn(),
+			action3: jest.fn(),
+		},
+		options: {},
+	};
+	let childProcessChannel;
```
Comment (shuse2) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: this let should be together with below let

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+
+	describe('#constructor', () => {
+		it('should create a local bus based on EventEmitter2', () => {
+			// Assert
+			expect(childProcessChannel.localBus).toBeInstanceOf(EventEmitter2);
+		});
+	});
+
+	describe('#registerToBus', () => {
+		beforeEach(() => childProcessChannel.registerToBus(socketsPath));
```
Comment (shuse2) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: prefer to consistently use 2 lines for test
```
beforeEach(async () => {
  childProcessChannel.registerToBus(socketsPath)
});
```

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+		pub: 'pub',
+		rpc: 'rpc',
+	};
+
+	let resolveWhenAllSocketsBound;
+	let rejectWhenAnySocketFailsToBind;
+	let rejectWhenTimeout;
+	let removeAllListeners;
+
+	beforeEach(() => {
```
Comment (shuse2) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: should we consistently use `async` for `it`, 'beforeEach` and `afterEach`?

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 5]
```diff
@@ -0,0 +1,622 @@
...
+		});
+	});
+
+	describe('#subscribe', () => {
+		const eventHandler = jest.fn();
+		const eventName = 'anEventName';
+
+		beforeEach(async () => {
+			await childProcessChannel.registerToBus(socketsPath);
+			return childProcessChannel.subscribe(eventName, eventHandler);
```
Comment (shuse2) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: return not needed

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 6]
```diff
@@ -0,0 +1,622 @@
...
+				// Assert
+				expect(error.message).toBe(
+					`Event "${eventName}" not registered in "differentModule" module.`
+				);
+			}
+		});
+	});
+
+	describe('#invoke', () => {
+		let ActionStub;
```
Comment (shuse2) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: prefer consistent `const` `let`ordering

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 7]
```diff
@@ -0,0 +1,622 @@
...
+						eventList,
+						actionsList,
+						socketOptions,
+						callback
+					) => {
+						callback(null, true);
+					}
+				);
+
+			return childProcessChannel._resolveWhenAllSocketsBound();
```
Comment (shuse2) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: no need to return

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js (hunk 3 above)
Comment (limiaspasdaniel) [reply framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: This has conflicts with `eslint` rule with `arrow-body-style`. I don't see the problem when there is only one line.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js (hunk 2 above)
Comment (limiaspasdaniel) [reply framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: Right, thanks for pointing out

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js (hunk 1 above)
Comment (limiaspasdaniel) [reply framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: Moved imports before mocking. But had to leave manual mocking of axon-RPC Client and Server as ChildProcessChannel needs to have those mocked when requiring them.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js (hunk 3 above)
Comment (shuse2) [reply framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: mmm i c, then it's ok.
In elements, we had this convention to always not to use one-liner, but i guess jest setup is different

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js (hunk 1 above)
Comment (shuse2) [reply framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: ah got it! 👍

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js (hunk 4 above)
Comment (limiaspasdaniel) [reply framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: I think there was an `eslint` rule for Mocha tests but not for Jest, that was forcing all the tests to be asynchronous.  I am not really sure if the same needs to be applied to `Jest`. Maybe @yatki can throw some light here. In my point of view, if a test doesn't contain asynchronous code, defining the function as async makes it misleading. This scenario is different from when we write `async` in a function definition so the user is aware that this function returns a promise.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js (hunk 5 above)
Comment (limiaspasdaniel) [reply framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: Right, thanks for pointing

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js (hunk 4 above)
Comment (shuse2) [reply framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: yea, I think that was my first thought too.
https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-synchronous-tests.md
but it's more like safety measure for unstable test

Comment (yatki) [reply framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: @shuse2 @limiaspasdaniel It's definitely not needed to use `async` keyword. 

- You need `async` when you need to await a promise.
- You need `done` when you need assertions inside a callback.

There is one important info that we all need to be careful that is using `resolves` / `rejects` keyword: https://jestjs.io/docs/en/expect#resolves

File: framework/src/controller/channels/child_process/index.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+const setupProcessHandlers = channel => {
+	process.once('SIGTERM', () => channel.cleanup(1));
+	process.once('SIGINT', () => channel.cleanup(1));
+	process.once('cleanup', (error, code) => channel.cleanup(code, error));
+	process.once('exit', (error, code) => channel.cleanup(code, error));
+};
+
+module.exports = {
+	setupProcessHandlers,
+};
```
Comment (yatki) [inline framework/src/controller/channels/child_process/index.js]: this file should have a test too.

File: framework/test/jest/specs/unit/controller/channels/in_memory_channel.spec.js [hunk 9]
```diff
@@ -14,10 +14,10 @@ describe('InMemoryChannel Channel', () => {
 			action2: jest.fn(),
 			action3: jest.fn(),
 		},
-		bus: new Bus(),
 		options: {},
 	};
 	let inMemoryChannel = null;
+	const bus = new Bus();
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/in_memory_channel.spec.js]: ❤️

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+jest.mock('axon-rpc', () => ({
+	Client: jest.fn(() => ({
+		call: jest.fn(),
+	})),
+	Server: jest.fn(() => ({
+		expose: jest.fn(),
+	})),
+}));
+
+jest.mock('axon');
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: please order mock functions from modules to files.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+	const socketsPath = {
+		root: 'root',
+		sub: 'sub',
+		pub: 'pub',
+		rpc: 'rpc',
+	};
+
+	let childProcessChannel;
+	let spies;
+	beforeEach(() => {
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: Please leave a blank line before hooks or test suites.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+			setupSockets: jest
+				.spyOn(childProcessChannel, 'setupSockets')
+				.mockResolvedValue(),
+		};
+	});
+
+	describe('inheritance', () => {
+		it('should be extended from BaseChannel class', () => {
+			// Assert
+			expect(Object.getPrototypeOf(childProcessChannel)).toBeInstanceOf(
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: No need for `Object.getPrototypeOf `.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+	describe('#constructor', () => {
+		it('should create a local bus based on EventEmitter2', () => {
+			// Assert
+			expect(childProcessChannel.localBus).toBeInstanceOf(EventEmitter2);
+		});
+	});
+
+	describe('#registerToBus', () => {
+		beforeEach(() => {
+			childProcessChannel.registerToBus(socketsPath);
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: Please add // Act comment

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+		it('should call setupSockets', () => {
+			// Assert
+			expect(spies.setupSockets).toHaveBeenCalledTimes(1);
+		});
+	});
+
+	describe('#subscribe', () => {
+		const validEventName = `${params.moduleAlias}:anEventName`;
+		beforeEach(async () => {
+			await childProcessChannel.registerToBus(socketsPath);
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: Please add Arrange comment

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+			);
+		});
+
+		it('should call setupSockets', () => {
+			// Assert
+			expect(spies.setupSockets).toHaveBeenCalledTimes(1);
+		});
+	});
+
+	describe('#subscribe', () => {
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: subscribe function should give a warning when event is not registered yet. So please add a todo test for that case. The same goes for publish and once.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+				invalidEventName,
+				expect.any(Function)
+			);
+		});
+	});
+
+	describe('#once', () => {
+		const validEventName = `${params.moduleAlias}:anEventName`;
+
+		beforeEach(() => childProcessChannel.registerToBus(socketsPath));
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: Please add an Arrange comment on top of beforeEach

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+		});
+	});
+
+	describe('#publish', () => {
+		const validEventName = `${params.moduleAlias}:anEventName`;
+		const invalidEventName = 'invalidModule:anEventName';
+
+		beforeEach(async () => {
+			// Arrange
+			await childProcessChannel.registerToBus(socketsPath);
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: remove async await and just return the method call

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 18]
```diff
@@ -0,0 +1,474 @@
...
+				params.actions,
+				{ skipInternalEvents: true }
+			);
+
+			jest
+				.spyOn(anotherChildProcessChannel, 'setupSockets')
+				.mockResolvedValue();
+
+			// Act
+			await anotherChildProcessChannel.registerToBus(socketsPath);
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: registerToBus should be in arrange block

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 19]
```diff
@@ -0,0 +1,474 @@
...
+			anotherChildProcessChannel.publish(validEventName, data);
+
+			// Assert
+			expect(anotherChildProcessChannel.pubSocket).toBe(undefined);
+		});
+	});
+
+	describe('#invoke', () => {
+		const actionName = 'moduleAlias:action1';
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: use parameter values from the `params` variable. For action1 as well.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 20]
```diff
@@ -0,0 +1,474 @@
...
+		});
+
+		it('should call setupSockets', () => {
+			// Assert
+			expect(spies.setupSockets).toHaveBeenCalledTimes(1);
+		});
+	});
+
+	describe('#subscribe', () => {
+		const validEventName = `${params.moduleAlias}:anEventName`;
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: Please get event name from `params.events[0]`. Apply the same for rest of the file.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 21]
```diff
@@ -0,0 +1,474 @@
...
+
+			// Act
+			await childProcessChannel.cleanup();
+
+			// Assert
+			expect(childProcessChannel.rpcSocket.close).toHaveBeenCalled();
+		});
+	});
+
+	describe('#_resolveWhenAllSocketsBound', () => {
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: Please use Arrange act comments in the test below.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 22]
```diff
@@ -0,0 +1,474 @@
...
+				childProcessChannel.eventsList.map(event => event.name),
+				childProcessChannel.actionsList.map(action => action.name),
+				{ type: 'ipcSocket', rpcSocketPath: childProcessChannel.rpcSocketPath },
+				expect.any(Function)
+			);
+		});
+	});
+
+	describe('#_rejectWhenAnySocketFailsToBind', () => {
+		beforeEach(() => childProcessChannel.registerToBus(socketsPath));
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: Please add arrange block.

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 23]
```diff
@@ -0,0 +1,474 @@
...
+					'error',
+					expect.any(Function)
+				);
+			}
+		});
+	});
+
+	describe('#_rejectWhenTimeout', () => {
+		beforeEach(async () => {
+			await childProcessChannel.registerToBus(socketsPath);
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: return method

File: framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js [hunk 24]
```diff
@@ -0,0 +1,474 @@
...
+		});
+
+		it('should reject with an Error object with proper message', async () => {
+			await expect(childProcessChannel._rejectWhenTimeout(1)).rejects.toThrow(
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/child_process_channel.spec.js]: return

File: framework/test/jest/specs/unit/controller/controller.spec.js [hunk 25]
```diff
@@ -77,9 +77,8 @@ describe('Controller Class', () => {
 				spies._setupDirectories
 			);
 			expect(spies._setupBus).toHaveBeenCalledAfter(spies._validatePidFile);
-			expect(spies._loadModules)
-				.toHaveBeenCalledAfter(spies._setupBus)
-				.toHaveBeenCalledWith(modules);
+			expect(spies._loadModules).toHaveBeenCalledAfter(spies._setupBus);
+			expect(spies._loadModules).toHaveBeenCalledWith(modules);
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/controller.spec.js]: @limiaspasdaniel this was solved with https://github.com/LiskHQ/lisk/pull/3179. Could you rebase your branch to development.

Comment (limiaspasdaniel) [reply framework/test/jest/specs/unit/controller/controller.spec.js]: I am not really sure of rebasing this branch to development. @lsilvs can it be done in this branch?

File: framework/src/controller/channels/child_process/index.js (hunk 8 above)
Comment (limiaspasdaniel) [reply framework/src/controller/channels/child_process/index.js]: Added skeleton for these tests and will be addressed in another issue.

File: framework/test/jest/specs/unit/controller/controller.spec.js (hunk 25 above)
Comment (lsilvs) [reply framework/test/jest/specs/unit/controller/controller.spec.js]: We will do before merging `feature/child_process_module` to development

**Pull request merged**